router/internal/expr/expr.go

```diff
@@ -0,0 +1,166 @@
+package expr
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"github.com/expr-lang/expr"
+	"github.com/expr-lang/expr/file"
+	"github.com/expr-lang/expr/vm"
+	"github.com/wundergraph/cosmo/router/pkg/authentication"
+	"net/http"
+	"net/url"
+	"reflect"
+)
+
+// RequestRootContext is the context for expressions parser when evaluating dynamic expressions
+// If possible function calls should be avoided in the expressions as they are much more expensive
+// See https://github.com/expr-lang/expr/issues/734
+type RequestRootContext struct {
+	Request RequestContext
+}
+
+// RequestContext is the context for the request object in expressions. Be aware, that only value receiver methods
+// are exported in the expr environment. This is because the expressions are evaluated in a read-only context.
+type RequestContext struct {
+	Auth   RequestAuth
+	Header RequestHeaders
+	URL    RequestURL
+}
+
+// RequestURL is the context for the URL object in expressions
+// it is limited in scope to the URL object and its components. For convenience, the query parameters are parsed.
+type RequestURL struct {
+	Method string
+	// Scheme is the scheme of the URL
+	Scheme string
+	// Host is the host of the URL
+	Host string
+	// Path is the path of the URL
+	Path string
+	// Query is the parsed query parameters
+	Query map[string]string
+}
+
+// RequestHeaders is the context for the headers object in expressions
+type RequestHeaders struct {
+	// Header is the http headers. We expose the header object to work with header case-insensitive
+	Header http.Header
+}
+
+// Get returns the first value associated with the given key (Exported).
+// For convenience, we export a function to make the work with headers case-insensitive.
+func (r RequestHeaders) Get(key string) string {
+	return r.Header.Get(key)
+}
+
+// LoadRequest loads the request object into the context.
+func (r *RequestRootContext) LoadRequest(req *http.Request) {
```

Comment:
This function is basically always called directly after creating a new context with 
`expr.RequestRootContext{}`

Maybe this could be a 
```go
NewFromRequest(req *http.Request) RequestRootContext {}
```

Reply:
We would still need to conditionally run `LoadAuth`. I'd then rather keep it consistent with `Load`.